doc/users/interactive_guide.rst

```diff
@@ -0,0 +1,432 @@
+handling UI events is useful, we can do better!  We really want to be
+able to have a usable prompt **and** interactive figure windows.
+
+We can do this using the 'input hook' feature of the interactive
+prompt.  This hook is called by the prompt as it waits for the user
+type (even for a fast typist the prompt is mostly waiting for the
+human to think and move their fingers).  Although the details vary
+between prompts the logic is roughly
+
+1. start to wait for keyboard input
+2. start the GUI event loop
+3. as soon as the user hits a key, exit the GUI event loop and handle the key
+4. repeat
+
+This gives us the illusion of simultaneously having an interactive GUI
+windows and an interactive prompt.  Most of the time the GUI event
+loop is running, but as soon as the user starts typing the prompt
+takes over again.
+
+This time-share technique only allows the event loop to run while
+python is otherwise idle and waiting for user input.  If you want the
+GUI to be responsive during long running code it is necessary to
+periodically flush the GUI event queue as described :ref:`above
+<spin_event_loop>`.  In this case it is your code, not the REPL, which
+is blocking process so you need to handle the "time-share" manually.
+Conversely, a very slow figure draw will block the prompt until it
+finishes drawing.
+
+Full embedding
+==============
+
+It is also possible to go the other direction and fully embed figures
+(and a `Python interpreter
+<https://docs.python.org/3/extending/embedding.html>`__) in a rich
+native application.  Matplotlib provides classes for each toolkit
+which can be directly embedded in GUI applications (this is how the
+built-in windows are implemented!).  See :ref:`user_interfaces` for
+more details.
+
+
+.. _interactive_scripts :
+
+Scripts
+=======
+
+
+.. autosummary::
+   :template: autosummary.rst
+   :nosignatures:
+
+   backend_bases.FigureCanvasBase.flush_events
+   backend_bases.FigureCanvasBase.draw_idle
+
+   figure.Figure.ginput
+   pyplot.ginput
+
+There are several use-cases for using interactive figures in scripts:
+
+- progress updates as a long running script progresses
+- capture user input to steer the script
+- streaming updates from a data source
+
+
+In the first if you only need to collect points in an Axes you can use
+`.figure.Figure.ginput` or more generally the tools from
+`.blocking_input` the tools will take care of starting and stopping
+the event loop for you.  However if you have written some custom event
+handling or are using `.widgets` you will need to manually run the GUI
+event loop using the methods described :ref:`above <cp_block_the_prompt>`.
+
+In the second caes, if you have open windows that have pending UI
+events (mouse clicks, button presses, or draws) you can explicitly
+process those events by calling `.FigureCanvasBase.flush_events`.
+This will run the GUI event loop until all UI events currently waiting
+have been processed.  The exact behavior is backend-dependent but
```

Comment:
is it?  should be be homogenized across backends?

Reply:
I am hedging here, but I think that where the variation, if there is any, will be is actually the the toolkit level not on our side.